### PR TITLE
Do not sell implementation

### PR DESCRIPTION
**Pull request by SlayterDev**
Task/Issue URL: https://app.asana.com/0/0/1191298152080292/f
Tech Design URL: https://app.asana.com/0/481882893211075/1195505979185439/f
CC: @brindy @catehstn @bwaresiak 

**Description**:
This implements the Do Not Sell (DN$) signal in the browser. There is a new setting labeled "Global Privacy Control" in the settings page. Turning this on will send the `Sec-GPC` header on main document requests to '1' and set the `navigator.globalPrivacyControl` property to `true`.

**Steps to test this PR**:
1. Navigate to http://global-privacy-control.glitch.me
2. Initially the header and DOM sections should say not present
3. Turn on the "Global Privacy Control" setting
4. The webpage will refresh and the header and DOM sections of the web page should say present

1. Navigate to another site like twitter.com
2. Manually inspect the webview via safari and check the header is included in the main document request and log the value of `navigator.globalPrivacyControl` to ensure it is `true`

**Copy Testing**:

* [x] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [x] iPhone SE (1st Gen)
* [x] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 11
* [x] iOS 12
* [x] iOS 13

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by SlayterDev** [comment]
While this is ready for review from a functionality stand point, the copy is subject to change.

File: DuckDuckGo/AppUserDefaults.swift
Comment (brindy) [inline DuckDuckGo/AppUserDefaults.swift]: Shouldn't need to use the Foundation namespace here?

File: DuckDuckGo/AppUserDefaults.swift
Comment (brindy) [inline DuckDuckGo/AppUserDefaults.swift]: We're moving to a model that uses `UserDefaultsWrapper`.  See `HomeMessageStorage` for a recent example.  I'll let you decide if you want to extract into a separate class, or leave it in here and use the wrapper.

While you're here - can you put a comment at the top of this class to tell future devs to look at `UserDefaultsWrapper`, please?  I should have done this a while ago really.  Thanks in advance!

File: DuckDuckGo/DoNotSellSettingsViewController.swift
Comment (brindy) [inline DuckDuckGo/DoNotSellSettingsViewController.swift]: Copyright header

File: DuckDuckGo/DoNotSellSettingsViewController.swift
Comment (brindy) [inline DuckDuckGo/DoNotSellSettingsViewController.swift]: We can't use hardcoded values here because this text will be translated at point (very soon). @bwaresiak is working on translations at the moment and will be able to give you an overview. 

In the mean time you can test this by going in to Xcode > Edit the scheme > Run > App Language and choose one of the pseudo languages at the bottom of the list.

File: DuckDuckGo/DoNotSellSettingsViewController.swift
Comment (brindy) [inline DuckDuckGo/DoNotSellSettingsViewController.swift]: This will need to be pulled in to `UserText` so that it can be translated.  Ping @bwaresiak for help as he's working on translations right now.

File: DuckDuckGo/DoNotSellSettingsViewController.swift
Comment (brindy) [inline DuckDuckGo/DoNotSellSettingsViewController.swift:29]: Table headers and footers have been ... problematic for the translations.  Once you fix the hardcoded string problem, can you please test with a pseudo language enabled (instructions in another comment) and check different screen sizes and especially rotation.  Again, please ping @bwaresiak for help here.

**Review by brindy — CHANGES_REQUESTED** [review]
Looking good so far, but there's a few things to address, mostly to do with supporting translations.   I haven't run it yet, either so please ping when they're resolved and the copy is final and I'll do the manual testing.  Thanks!

File: DuckDuckGo/AppUserDefaults.swift
Comment (brindy) [reply DuckDuckGo/AppUserDefaults.swift]: Also @samsymons is doing something with notifications to trigger a refresh when the content blocker data changes.  I think you can probably share the same concept.  Probably worth having a chat.

Comment (SlayterDev) [reply DuckDuckGo/AppUserDefaults.swift]: Looking at @samsymons PR I'm not sure where this would benefit this case although maybe I'm missing it. Since this setting is independent and does not relate to content blocking I feel having it as its own notification makes sense. I'm open to other opinions though.

File: DuckDuckGo/DoNotSellSettingsViewController.swift
Comment (SlayterDev) [reply DuckDuckGo/DoNotSellSettingsViewController.swift:29]: Tested this with a few different pseudo languages and appears to be displaying properly.

**Comment by SlayterDev** [comment]
FYI copy has been finalized

File: DuckDuckGo/TabViewController.swift
Comment (brindy) [inline DuckDuckGo/TabViewController.swift:974]: Can you extract this in to a private function, please? 

This chunk of code could be replaced like this:
```
    if let request = requestForDoNotSell(basedOn: navigationAction.request) {
        decisionHandler(.cancel)
        load(urlRequest: request)
        return
    }
```

With a private function like this (untested btw):
```
private func requestForNotSell(basedOn incomingRequest: request) -> URLRequest? {

     var request = incomingRequest
     if appSettings.sendDoNotSell {
             if let headers = request.allHTTPHeaderFields,
                headers.firstIndex(where: { $0.key == "Sec-GPC" }) == nil {
                 request.addValue("1", forHTTPHeaderField: "Sec-GPC")
                 return request
             }
         } else {
             // Check if DN$ header is still there and remove it
             if let headers = request.allHTTPHeaderFields,
                let _ = headers.firstIndex(where: { $0.key == "Sec-GPC" }) {
                 request.setValue(nil, forHTTPHeaderField: "Sec-GPC")
                 return request
             }
         }

    return nil
}
```

File: DuckDuckGo/TabViewController.swift
Comment (brindy) [inline DuckDuckGo/TabViewController.swift]: These "Sec-GPC" strings, can you add them to the `Constants` struct at the top of the class, please?  Then use like:

`headers.firstIndex(where: { $0.key == Constants.doNotSellKey }) == nil {`

(Whatever name is most appropriate)

File: DuckDuckGo/DoNotSellSettingsViewController.swift
Comment (brindy) [inline DuckDuckGo/DoNotSellSettingsViewController.swift:37]: In general all the blurb text's style doesn't match what is in Figma - if I'm looking at the right [place](https://www.figma.com/file/g1Ucu8mVgtHfVrNzrKyKUK/Global-Privacy-Control--Apps-and-Extensions?node-id=214%3A1294).  Font size, color and line spacing all appear to be off.  

In Figma you can get some tips about line spacing by clicking the text and choosing  iOS from the code drop down.  

You may need to work backwards from Figma to determine which built in color to use (shouldn't be adding any new ones), though you may find it makes sense to add a couple of new theme attributes.

Once you've made the changes, a matter of process (which tbf you didn't know about), can you share some screenshots in both light and dark mode with Mike in a sub task and get him to give them a thumbs up, please?

File: DuckDuckGo/UserText.swift
Comment (brindy) [inline DuckDuckGo/UserText.swift]: There doesn't appear to be an indent in the designs in Figma.

File: DuckDuckGo/UserText.swift
Comment (brindy) [inline DuckDuckGo/UserText.swift]: @bwaresiak - is it OK to use multiline text like this from the translation perspective, or should it use new line characters?

**Review by brindy — CHANGES_REQUESTED** [review]
Hey @SlayterDev - good stuff, a few more changes though I'm afraid.  Give me a ping if you want to chat through any of it.

File: DuckDuckGo/UserText.swift
Comment (bwaresiak) [reply DuckDuckGo/UserText.swift]: That's a good question - I haven't tested that. 
@SlayterDev Can you / Did you regenerate Localizable.strings file? You can do that by running command `find DuckDuckGo/ -name "*.swift" -print0 | xargs -0 xcrun extractLocStrings -o DuckDuckGo/en.lproj` (run in the main project dir) - this is automated on my branch but still as work in progress. Sorry for the pain!

**Review by brindy — APPROVED** [review]
LGTM!  As long as it's only minor UI changes come out of the Product Review, if anything, feel free to merge (once the Product Review is approved).  Thanks for picking this up!

**Pull request merged**
